internal/controllers/machinedeployment/machinedeployment_controller.go

```diff
@@ -238,16 +246,73 @@ func (r *Reconciler) reconcile(ctx context.Context, cluster *clusterv1.Cluster,
 		if d.Spec.Strategy.RollingUpdate == nil {
 			return ctrl.Result{}, errors.Errorf("missing MachineDeployment settings for strategy type: %s", d.Spec.Strategy.Type)
 		}
-		return ctrl.Result{}, r.rolloutRolling(ctx, d, msList)
+		return ctrl.Result{}, r.rolloutRolling(ctx, d, machineSets)
 	}
 
 	if d.Spec.Strategy.Type == clusterv1.OnDeleteMachineDeploymentStrategyType {
-		return ctrl.Result{}, r.rolloutOnDelete(ctx, d, msList)
+		return ctrl.Result{}, r.rolloutOnDelete(ctx, d, machineSets)
 	}
 
 	return ctrl.Result{}, errors.Errorf("unexpected deployment strategy type: %s", d.Spec.Strategy.Type)
 }
 
+func (r Reconciler) getMachinesForMachineDeployment(ctx context.Context, machineDeployment *clusterv1.MachineDeployment) ([]*clusterv1.Machine, error) {
+	var machineList clusterv1.MachineList
+	if err := r.Client.List(
+		ctx,
+		&machineList,
+		client.MatchingLabels(machineDeployment.Spec.Selector.MatchLabels),
+		client.InNamespace(machineDeployment.GetNamespace()),
+	); err != nil {
+		return nil, err
+	}
+
+	machines := make([]*clusterv1.Machine, 0)
+	for key, _ := range machineList.Items {
+		machines = append(machines, &machineList.Items[key])
+	}
+
+	return machines, nil
+}
+
+func (r Reconciler) syncManagedLabelsInPlace(ctx context.Context, machineSets []*clusterv1.MachineSet, machines []*clusterv1.Machine, desiredNodeLabels *map[string]string) error {
```

Comment:
q: (not for this PR, but for a follow up) is this the place to take care of other in-place changes propagation like the one suggested in https://github.com/kubernetes-sigs/cluster-api/issues/5880